datadog/client.go

```diff
@@ -1,103 +1,176 @@
-	} else {
-		log.Printf("stats/datadog: connection opened to %s with a buffer size of %d B", config.Address, cap(conn.b))
+	if config.BufferSize == 0 {
+		config.BufferSize = DefaultBufferSize
 	}
 
-	return &Client{
-		conn: conn,
+	c := &Client{
+		serializer: serializer{},
+	}
+
+	conn, bufferSize, err := dial(config.Address, config.BufferSize)
+	if err != nil {
+		log.Printf("stats/datadog: %s", err)
 	}
+
+	c.conn, c.err = conn, err
+	c.buffer.BufferSize = bufferSize
+	c.buffer.Serializer = &c.serializer
+	return c
 }
 
-// Close satisfies the io.Closer interface.
-func (c *Client) Close() (err error) {
-	c.once.Do(func() {
-		if c.conn != nil {
-			err = c.conn.Close()
-		}
-	})
-	return
+// HandleMetric satisfies the stats.Handler interface.
+func (c *Client) HandleMeasures(time time.Time, measures ...stats.Measure) {
+	c.buffer.HandleMeasures(time, measures...)
 }
 
-// Flsuh satisfies the stats.Flusher interface.
+// Flush satisfies the stats.Flusher interface.
 func (c *Client) Flush() {
-	if c.conn != nil {
-		if err := c.conn.Flush(); err != nil {
-			log.Printf("stats/datadog: sending metrics to %s failed: %s", c.conn.RemoteAddr(), err)
-		}
+	c.buffer.Flush()
+}
+
+// Close flushes and closes the client, satisfies the io.Closer interface.
+func (c *Client) Close() error {
+	c.Flush()
+	c.close()
+	return c.err
+}
+
+type serializer struct {
+	conn net.Conn
+}
+
+func (*serializer) AppendMeasures(b []byte, _ time.Time, measures ...stats.Measure) []byte {
+	for _, m := range measures {
+		b = AppendMeasure(b, m)
 	}
+	return b
 }
 
-// HandleMetric satisfies the stats.Handler interface.
-func (c *Client) HandleMetric(m *stats.Metric) {
-	if c.conn != nil {
-		buf := bufferPool.Get().(*buffer)
-		buf.b = appendMetric(buf.b[:0], Metric{
-			Type:      metricType(m),
-			Namespace: m.Namespace,
-			Name:      m.Name,
-			Value:     m.Value,
-			Tags:      m.Tags,
-		})
-		if _, err := c.conn.Write(buf.b); err != nil {
-			log.Printf("stats/datadog: sending metric %s to %s failed: %s", m.Name, c.conn.RemoteAddr(), err)
+func (s *serializer) Write(b []byte) (int, error) {
+	if s.conn != nil {
+		return s.conn.Write(b)
+	}
+	return 0, io.ErrClosedPipe
+}
+
+func (s *serializer) close() {
+	if s.conn != nil {
+		s.conn.Close()
+	}
+}
+
+func dial(address string, sizehint int) (conn net.Conn, bufsize int, err error) {
+	var f *os.File
+
+	if conn, err = net.Dial("udp", address); err != nil {
+		return
+	}
+
+	if f, err = conn.(*net.UDPConn).File(); err != nil {
+		conn.Close()
+		return
+	}
+	defer f.Close()
+	fd := int(f.Fd())
+
+	// The kernel refuses to send UDP datagrams that are larger than the size of
+	// the size of the socket send buffer. To maximize the number of metrics
+	// sent in one batch we attempt to attempt to adjust the kernel buffer size
+	// to accept larger datagrams, or fallback to the default socket buffer size
+	// if it failed.
```

Comment:
Those are limitations of both the UDP protocol and its implementation in the kernel. But worse, it also has to fit in the receiver's buffer, otherwise the datagram would get truncated on read as well, and the sender side has no way to know how big the receiver's buffer is.

We've had a couple of performance issues due to small UDP buffers, the datadog agent supported only 1KB datagrams for long, I think they raised it to 4KB or 8KB (which is still very small). This is one of the reasons we moved to using veneur, which we configure to use a 32KB read buffer.

They say UDP stands for User Datagram Protocol, but it feels more like Unreliable Datagram Protocol ;)